tools/xkbcli-completion.sh

```diff
@@ -0,0 +1,97 @@
+# bash completion support for xkbcli.
+
+# See completion API documentation: https://github.com/scop/bash-completion
+# NOTE: The script parses the commands help messages to provide the completions,
+#       thus any new subcommand or option will be supported, as long as it has its
+#       entry in the help messages. This should result in low maintenancei effort.
+
+___xkbcli_main()
+{
+    # Initialization: https://github.com/scop/bash-completion/blob/fdf4456186eb4548ef628e65fb1be73d8e4695e9/bash_completion.d/000_bash_completion_compat.bash#L205
+    local cur prev words cword
+    _init_completion -s || return
+
+    local i=1 cmd
+
+    # Find subcommand
+    while [[ "$i" -lt "$COMP_CWORD" ]]
+    do
+        local s="${COMP_WORDS[i]}"
+        case "$s" in
+            -*) ;;
+            *)
+            cmd="$s"
+            break
+            ;;
+        esac
+        (( i++ ))
+    done
+
+    # Parse available subcommands
+    local help_output line is_command_list=false subcommands=()
+    help_output=$(xkbcli --help)
+    while IFS= read -r line; do
+        # Traverse subcommand list
+        if [[ "$is_command_list" == true ]]; then
+            # Check for subcommand based on the indentation
+            if [[ "$line" =~ ^[[:blank:]]{2}([[:alpha:]]([[:alnum:]]|-)+)$ ]]; then
+                subcommands+=("${BASH_REMATCH[1]}")
+            # Detect end of subcommand list based on indentation
+            elif [[ "$line" =~ ^[[:graph:]] ]]; then
+                is_command_list=false
+            fi
+        # Detect start of subcommand list
+        elif [[ "$line" == "Commands:" ]]; then
+            is_command_list=true
+        fi
+    done <<< "$help_output"
+
+    # No previous subcommand or incomplete: completion for root xkbcli command
+    if [[ "$i" -eq "$COMP_CWORD" ]]
+    then
+        local opts
+        # Doc for _parse_help: https://github.com/scop/bash-completion/blob/fdf4456186eb4548ef628e65fb1be73d8e4695e9/bash_completion.d/000_bash_completion_compat.bash#L311
+        opts=$(_parse_help xkbcli)
+        local cur="${COMP_WORDS[COMP_CWORD]}"
+        COMPREPLY=($(compgen -W "${subcommands[*]} $opts" -- "$cur"))
+        return
+    fi
+
+    # Found a supported subcommand: proceed to completion
+    if [[ " ${subcommands[*]} " =~ " $cmd " ]]
```

Comment:
are the spaces here on purpose?

also, `if [[ ]]; then` for consistency please

Reply:
> are the spaces here on purpose?

I admit I adapted a copy/paste and did not pay much attention because it worked 😅. So the spaces are needed because we need to match _whole_ subcommands. `${subcommands[*]}` provide a space-separated list, but we need space before and after.

I switched to an explicit pattern. There may be better solution than pattern matching though.